corehq/apps/translations/app_translations/utils.py

```diff
@@ -0,0 +1,226 @@
+# coding=utf-8
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import six
+from django.contrib import messages
+from django.utils.translation import ugettext as _
+
+from corehq.apps.app_manager.const import APP_TRANSLATION_UPLOAD_FAIL_MESSAGE
+from corehq.util.python_compatibility import soft_assert_type_text
+from corehq.apps.translations.const import (
+    LEGACY_MODULES_AND_FORMS_SHEET_NAME,
+    MODULES_AND_FORMS_SHEET_NAME,
+    SINGLE_SHEET_NAME,
+)
+from corehq.util.workbook_json.excel import HeaderValueError, WorkbookJSONReader, JSONReaderError, \
+    InvalidExcelFileException
+
+
+def get_app_translation_workbook(file_or_filename):
+    msgs = []
+    try:
+        workbook = WorkbookJSONReader(file_or_filename)
+    # todo: HeaderValueError does not belong here
+    except (HeaderValueError, InvalidExcelFileException) as e:
+        msgs.append(
+            (messages.error, _(APP_TRANSLATION_UPLOAD_FAIL_MESSAGE).format(e))
+        )
+        return False, msgs
+    except JSONReaderError as e:
+        msgs.append(
+            (messages.error, _(
+                "App Translation Failed! There is an issue with Excel columns. Error details: {}."
+            ).format(e))
+        )
+        return False, msgs
+    return workbook, msgs
+
+
+def get_bulk_app_sheet_headers(app, lang=None, exclude_module=None, exclude_form=None):
+    '''
+    Returns lists representing the expected structure of bulk app translation
+    Excel file uploads and downloads.
+
+    The list will be in the form:
+    [
+        ["sheetname", ["column name1", "column name 2"]],
+        ["sheet2 name", [...]],
+        ...
+    ]
+
+    exclude_module and exclude_form are functions that take in one argument
+    (form or module) and return True if the module/form should be excluded
+    from the returned list
+    '''
+    langs = [lang] if lang else app.langs
+
+    default_lang_list = ['default_' + l for l in langs]
+    audio_lang_list = ['audio_' + l for l in langs]
+    image_lang_list = ['image_' + l for l in langs]
+    video_lang_list = ['video_' + l for l in langs]
+    lang_list = default_lang_list + image_lang_list + audio_lang_list + video_lang_list
+
+    if lang:
+        return ((SINGLE_SHEET_NAME, (
+            'menu_or_form',
+            'case_property',        # modules only
+            'list_or_detail',       # modules only
+            'label',                # forms only
+        ) + tuple(lang_list)),)
+
+    headers = []
+
+    # Add headers for the first sheet
+    headers.append([
+        MODULES_AND_FORMS_SHEET_NAME,
+        get_modules_and_forms_row(
+            row_type='Type',
+            sheet_name='menu_or_form',
```

Comment:
trying to convince you for a better name here if possible.
menu_or_form would have values like `module1` and `module1_form2` right?
`menu_or_form` sounds like this would have values ['module', 'form']. I would push to rename this back to `sheet_name` only or `module_or_form_name` instead.

Reply:
I can change to `module_or_form_name`. I'd rather not change back to `sheet_name` because that doesn't make sense for the single-sheet format, where these strings aren't sheet names.

Reply:
Rather, I'll change to `menu_or_form_name`.

Reply:
...although I just realized that `menu_or_form_name` sounds like that column should contain the name, not a string like `menu1`. It'd be more accurate to use `menu_or_form_slug`, but `slug` isn't a user-friendly name. So I agree that `menu_or_form` sounds like values would be either `menu` or `form` but I don't have a better name. Could do something really generic, like `item`, but that isn't really clear either.